### DISCUSSION TITLE
`.r-stretch` did not work in reveal.js presentation in some situations

### DISCUSSION THREAD

**Discussion by ttzytt** [quarto-dev/quarto-cli#7774] [Q&A]
### Description

Sometimes, it will be helpful to use `.r-stretch` to fit more complicated elements than images. For example: 

```qmd 
---
format: revealjs
---

abc abc abc 

abc abc 

abc abc

abc abc


::: {.r-stretch layout="[[-30, 20, 20, -30], [-25, 50, -25]]"}
![](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6a/Black_Color.jpg/675px-Black_Color.jpg)

![](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6a/Black_Color.jpg/675px-Black_Color.jpg?20210612095317)

![](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6a/Black_Color.jpg/675px-Black_Color.jpg?20210612095317)
::: 
```

But it seems like `.r-stretch` is not working in this case, as the images overflowed: 

![image](https://github.com/quarto-dev/quarto-cli/assets/92229358/1b5b2bdb-8c1e-4d12-b625-b4159282153a)

I'm not sure if this is a bug or if this is because I'm not using Quarto properly. Please provide some suggestions.

**Comment by ttzytt** [comment]
There is another strange problem relating to `.r-stretch`: 

```qmd
---
format: revealjs
---

## 

abc abc 

abc abc 

abc abc 

abc abc 


::: {.r-stretch .fragment .fade-up}

![](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6a/Black_Color.jpg/675px-Black_Color.jpg)

:::

## 

abc abc 

abc abc 

abc abc 

abc abc 

![](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6a/Black_Color.jpg/675px-Black_Color.jpg){.r-stretch .fragment .fade-up}

```

Theoretically, the two slides should be equivalent, but it turns out that the image is not automatically stretched to fit the rest of the space in the first slide, but it did work in the second.

First slide: 

![image](https://github.com/quarto-dev/quarto-cli/assets/92229358/de21ea3f-6888-40b9-8dff-8451b67cf217)

Second slide:
![image](https://github.com/quarto-dev/quarto-cli/assets/92229358/1180615f-bbd9-4bb9-b38c-68b6f7bd32b0)

**Reply by cderv** [reply]
> Theoretically, the two slides should be equivalent, but it turns out that the image is not automatically stretched to fit the rest of the space in the first slide, but it did work in the second.

In the first case the `.r-stretch` is applied on a `div` HTML element
In the second case, it is applied on the `img` HTML element. 

[Stretch is a Revealjs Feature](https://github.com/quarto-dev/quarto-cli/discussions/7774#discussioncomment-7751413). So not exactly the same. The sizing computation is done by Revealjs JS code directly, and I don't know how stretching works with Fragment. Maybe there is limitation.

**Comment by cderv** [comment]
As a clarification for context about the `.r-stretch` class: this is a feature supported by Revealjs directly: https://revealjs.com/layout/#stretch with [limitations](https://revealjs.com/layout/#stretch-limitations)

This is a way to autosize some HTML elements based other content before. Revealjs will run some JS code to compute size dynamically for the element. 

In your example, `stretch` is applying on the div, so I believe the sizing applied on the div will be ok. If you want to apply to `img` inside to auto scale, this seems to be different, and I am not sure RevealJS supports that.

**Reply by ttzytt** [reply]
Thank you very much for responding quickly, but I still have a few questions.

According to the limitations section in the Revealjs documentation you sent, it seems like the only limitation is the following two: 

> - Only direct descendants of a slide section can be stretched
> - Only one descendant per slide section can be stretched

Is the reason the first slide is not working because the image enclosed by the `div` element is not the "direct descendant" of a slide section? 

If the `div` element is indeed stretched properly, then shouldn't the image elements inside it not exceed the `div`?

Just to make sure, I created another example where the image element should take up the entire `div`, but `.r-stretch` is still not working: 

```qmd
---
format: revealjs
---

## 

abc abc 

abc abc 

abc abc 

abc abc 


::: {.r-stretch .fragment .fade-up}

![](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6a/Black_Color.jpg/675px-Black_Color.jpg){style="width: 100%; height: auto;"}

:::
```

It seems like the whole thing is meaningless when you can properly stretch the image if I do it in the way presented by the second slide. But the main problem here is that this is not possible if I want to properly present more complicated images, like the slide presented in the original post: 

```qmd
---
format: revealjs
---

abc abc abc 

abc abc 

abc abc

abc abc


::: {.r-stretch layout="[[-30, 20, 20, -30], [-25, 50, -25]]"}
![](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6a/Black_Color.jpg/675px-Black_Color.jpg)

![](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6a/Black_Color.jpg/675px-Black_Color.jpg?20210612095317)

![](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6a/Black_Color.jpg/675px-Black_Color.jpg?20210612095317)
::: 
```

**Reply by cderv** [reply]
I am not sure what is the expected results you are seeking as you are mixing several features together 

* Quarto `layout` for several images 
* Revealjs auto-sizing computation with `.r-stretch`
* Revealjs `fragment` that does hide some content to reveal it later

As I said, I am not sure if stretching and fragmented revealjs feature are correctly designed to work well together (I am thinking about te auto size computation that needs to know about the fragmented element that will be revealed later)

For example Quarto is not applying auto-stretch inside fragment div for example. So I would expect applying it to the div does not help either. (more on auto-stretch scope limitation at https://github.com/quarto-dev/quarto-cli/pull/896 for reference)

Initially, I believe Revealjs has support of the feature to size up to remaining space the content of an image.

<table>
<tr> 
 <td>no stretch
 <td>with stretch
<tr>
 <td>

````markdown
---
format: revealjs
---

## {.nostretch}

abc abc 

abc abc 

abc abc 

abc abc 


![](https://placehold.co/100)

````

 <td>

````markdown
---
format: revealjs
---

## 

abc abc 

abc abc 

abc abc 

abc abc 


![](https://placehold.co/100)
````

<tr>
 <td>

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/505fe094-3fb6-459e-8982-f22643ebf575)

 <td>

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/266f78e5-e304-4534-9983-3b6e6945a2b0)
</table>

When using this feature for more complex sizing, I am not sure it is the right fit. 

And looking at our code, I don't think setting `.r-stretch` on a div is supposed to work. Where did you see this usage ? 

Our doc mentions auto stretching feature and applying on image individually when auto-stretch is false 
https://quarto.org/docs/presentations/revealjs/advanced.html#stretch

**Reply by ttzytt** [reply]
I guess I misunderstood the correct usage of `.r-stretch`. So what I'm expecting here is that using my code in my initial post, the three images inside the `layout` will not overflow.  

From your responses, it seems like the `.r-stretch` is designed to work only for simple images. Then I wonder what will be the best practice for re-sizing a complicated element, say, the layout, to fill the rest of the space? 

If there is no easy way, then I guess this Q&A discussion page could be turned into a feature request of a more universal `.r-stretch` that can support all kinds of elements. (Although I'm not sure if this is the right place to ask for this feature as you've mentioned that `.r-stretch` directly came from Revealjs. But if in native Revealjs, the `.r-stretch` is universally applicable to all elements, this will be a nice and helpful feature to add in Quarto).
